packages/flutter_tools/gradle/src/main/kotlin/FlutterPluginUtils.kt

```diff
@@ -371,4 +392,433 @@ object FlutterPluginUtils {
+            // TODO(gmackall): This is doing a string comparison, which is odd and also can be wrong
+            //                 when comparing preview versions (against non preview, and also in the
+            //                 case of alphabet reset which happened with "Baklava".
+            if (pluginCompileSdkVersion > projectCompileSdkVersion) {
+                project.logger.quiet("Warning: The plugin $pluginName requires Android SDK version $pluginCompileSdkVersion or higher.")
+                project.logger.quiet(
+                    "For more information about build configuration, see ${FlutterPluginConstants.WEBSITE_DEPLOYMENT_ANDROID_BUILD_CONFIG}."
+                )
+            }
+
+            getAndroidExtension(project).buildTypes.forEach { buildType ->
+                addEmbeddingDependencyToPlugin(project, pluginProject, buildType, engineVersion)
+            }
+        }
+    }
+
+    private fun addEmbeddingDependencyToPlugin(
+        project: Project,
+        pluginProject: Project,
+        buildType: BuildType,
+        engineVersion: String
+    ) {
+        val flutterBuildMode: String = buildModeFor(buildType)
+        // In AGP 3.5, the embedding must be added as an API implementation,
+        // so java8 features are desugared against the runtime classpath.
+        // For more, see https://github.com/flutter/flutter/issues/40126
+        if (!supportsBuildMode(pluginProject, flutterBuildMode)) {
+            return
+        }
+        if (!pluginProject.hasProperty("android")) {
+            return
+        }
+
+        // Copy build types from the app to the plugin.
+        // This allows to build apps with plugins and custom build types or flavors.
+        getAndroidExtension(pluginProject).buildTypes.addAll(getAndroidExtension(project).buildTypes)
+
+        // The embedding is API dependency of the plugin, so the AGP is able to desugar
+        // default method implementations when the interface is implemented by a plugin.
+        //
+        // See https://issuetracker.google.com/139821726, and
+        // https://github.com/flutter/flutter/issues/72185 for more details.
+        addApiDependencies(pluginProject, buildType.name, "io.flutter:flutter_embedding_$flutterBuildMode:$engineVersion")
+    }
+
+    // ------------------ Task adders (a subset of the above category)
+
+    // Add a task that can be called on flutter projects that prints the Java version used in Gradle.
+    //
+    // Format of the output of this task can be used in debugging what version of Java Gradle is using.
+    // Not recommended for use in time sensitive commands like `flutter run` or `flutter build` as
+    // Gradle is slower than we want. Particularly in light of https://github.com/flutter/flutter/issues/119196.
+    @JvmStatic
+    @JvmName("addTaskForJavaVersion")
+    internal fun addTaskForJavaVersion(project: Project) {
+        project.tasks.register("javaVersion") {
```

Comment:
Unrelated to this pr we should figure out how gradlew tasks works and add our own section with targets that might make sense for flutter devs (like this one)